src/main/scala/higherkindness/compendium/core/ProtocolUtils.scala

```diff
@@ -16,28 +16,75 @@
 
 package higherkindness.compendium.core
 
-import cats.effect.Sync
+import java.io.{File, PrintWriter}
+
+import cats.effect.{Resource, Sync}
 import cats.implicits._
-import higherkindness.compendium.models.Protocol
+import higherkindness.compendium.models.{IdlName, Protocol}
 import higherkindness.compendium.models.transformer.types.SchemaParseException
+import higherkindness.droste.data.Mu
+import higherkindness.skeuomorph.protobuf.{ProtobufF, Protocol => ProtobufProtocol}
+import higherkindness.skeuomorph.protobuf.ParseProto._
 import org.apache.avro.Schema
 
 trait ProtocolUtils[F[_]] {
-  def validateProtocol(protocol: Protocol): F[Protocol]
+  def validateProtocol(protocol: Protocol, schema: IdlName): F[Protocol]
 }
 
 object ProtocolUtils {
 
-  private def parser: Schema.Parser = new Schema.Parser()
+  private def parserAvro: Schema.Parser = new Schema.Parser()
+
+  private def parserProtobuf[F[_]: Sync](raw: String): F[ProtobufProtocol[Mu[ProtobufF]]] = {
+    for {
+      tmpFile <- writeTempFile(raw)
+      p <- parseProto[F, Mu[ProtobufF]].parse(
+        ProtoSource(tmpFile.getName, tmpFile.getAbsolutePath.replaceAll(tmpFile.getName, ""))
+      )
+    } yield p
+  }
+
+  private def writeTempFile[F[_]: Sync](msg: String): F[File] = {
+    Resource
+      .make(F.delay {
+        val file = File.createTempFile("protoTempFile", ".proto")
+        file.deleteOnExit
+        file
+      })({ file: File =>
+        F.delay {
+          val pw = new PrintWriter(file)
+          pw.write(msg)
+          pw.close()
+        }
+      })
+      .use((file: File) => F.delay(file))
```

Comment:
I'm not sure if you are using the acquire and release branches correctly.

Probably the next code isn't perfect but the `release` part should be exclusively for closing the opened things safely:

```scala
final case class FilePrintWriter(file: File, pw: PrintWriter)

  private def parserAvro: Schema.Parser = new Schema.Parser()

  private def parserProtobuf[F[_]: Sync](raw: String): F[ProtobufProtocol[Mu[ProtobufF]]] = {
    for {
      fpw <- writeTempFile(raw)
      tmpFile = fpw.file
      p <- parseProto[F, Mu[ProtobufF]].parse(
        ProtoSource(tmpFile.getName, tmpFile.getAbsolutePath.replaceAll(tmpFile.getName, ""))
      )
    } yield p
  }

  private def writeTempFile[F[_]: Sync](msg: String): F[FilePrintWriter] = {
    Resource
      .make(F.delay {
        val file = File.createTempFile("protoTempFile", ".proto")
        file.deleteOnExit()
        FilePrintWriter(file, new PrintWriter(file))
      }) { fpw: FilePrintWriter =>
        F.delay(fpw.pw.close())
      }
      .use((fpw: FilePrintWriter) => F.delay(fpw.pw.write(msg)).as(fpw))
  }
```

Reply:
Yes that makes a lot of sense to me, I was having problems with retrieving  the `File` but that solves it